File: app/src/main/java/com/kjjejones42/audiobookplayer/player/PlayerNotificationManager.java
package com.kjjejones42.audiobookplayer.player;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import androidx.core.app.NotificationCompat;
import androidx.media.session.MediaButtonReceiver;

import com.kjjejones42.audiobookplayer.R;

public class PlayerNotificationManager {
    static final private String CHANNEL_ID = "com.kjjejones42.audiobookplayer";

    private final NotificationCompat.Builder notificationBuilder;

    private final MediaSessionCompat mediaSession;

    private final Context context;

    public PlayerNotificationManager(MediaSessionCompat mediaSession, Context context) {
        this.mediaSession = mediaSession;
        this.context = context;
        notificationBuilder = new NotificationCompat.Builder(context, CHANNEL_ID);
        initializeNotification();
    }

    private void initializeNotification() {
        CharSequence name = context.getString(R.string.channel_name);
        int importance = NotificationManager.IMPORTANCE_LOW;
        NotificationChannel channel = new NotificationChannel(CHANNEL_ID, name, importance);
        channel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
        context.getSystemService(NotificationManager.class).createNotificationChannel(channel);
        notificationBuilder
                .setContentIntent(mediaSession.getController().getSessionActivity())
                .setDeleteIntent(MediaButtonReceiver.buildMediaButtonPendingIntent(context,
                        PlaybackStateCompat.ACTION_STOP))
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setSmallIcon(R.drawable.ic_logo)
                .setShowWhen(false)
                .setStyle(new androidx.media.app.NotificationCompat.MediaStyle()
                        .setMediaSession(mediaSession.getSessionToken())
                        .setShowActionsInCompactView(1, 2, 3)
                        .setShowCancelButton(true)
                        .setCancelButtonIntent(MediaButtonReceiver.buildMediaButtonPendingIntent(context,
                                PlaybackStateCompat.ACTION_STOP)));
        setActions(false);
    }

    private void setActions(boolean playing) {
        int playPauseIcon = playing ? R.drawable.ic_pause : R.drawable.ic_play;
        String playPauseText = playing ? "Pause" : "Play";
        notificationBuilder
                .clearActions()
                .addAction(R.drawable.ic_skip_prev,
                        "Prev", MediaButtonReceiver.buildMediaButtonPendingIntent(
                        context, PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS))
                .addAction(R.drawable.ic_replay_30,
                        "Rewind", MediaButtonReceiver.buildMediaButtonPendingIntent(
                        context, PlaybackStateCompat.ACTION_REWIND))
                .addAction(
                        playPauseIcon, playPauseText,
                        MediaButtonReceiver.buildMediaButtonPendingIntent(context,
                                PlaybackStateCompat.ACTION_PLAY_PAUSE))
                .addAction(R.drawable.ic_forward_30,
                        "Forward", MediaButtonReceiver.buildMediaButtonPendingIntent(
                        context, PlaybackStateCompat.ACTION_FAST_FORWARD))
                .addAction(R.drawable.ic_skip_next,
                        "Next", MediaButtonReceiver.buildMediaButtonPendingIntent(
                        context, PlaybackStateCompat.ACTION_SKIP_TO_NEXT));

    }

    public Notification updateNotification(boolean playing, String bookName) {

        MediaControllerCompat controller = mediaSession.getController();

        MediaDescriptionCompat description = controller.getMetadata().getDescription();

        notificationBuilder
                .setContentText(description.getTitle())
                .setLargeIcon(description.getIconBitmap())
                .setOngoing(playing)
                .setContentTitle(bookName);
        setActions(playing);
        return notificationBuilder.build();
    }

}

File: app/src/main/java/com/kjjejones42/audiobookplayer/player/PlayerNotificationManager.kt
package com.kjjejones42.audiobookplayer.player

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.media.app.NotificationCompat
import androidx.media.session.MediaButtonReceiver
import com.kjjejones42.audiobookplayer.R
import com.kjjejones42.audiobookplayer.display.DisplayListActivity

private const val CHANNEL_ID = "com.kjjejones42.audiobookplayer"

class PlayerNotificationManager(
    private val mediaSession: MediaSessionCompat,
    private val context: Context
) {


    init {
        initializeNotification(context)
    }

    private fun initializeNotification(context: Context) {
        val name: CharSequence = context.getString(R.string.channel_name)
        val importance = NotificationManager.IMPORTANCE_LOW
        val channel = NotificationChannel(CHANNEL_ID, name, importance)
        channel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
        context.getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
    }

    fun updateNotification(playing: Boolean, bookName: String?): Notification {
        val controller = mediaSession.controller

        val description = controller.metadata.description

        val intent = Intent(context, PlayActivity::class.java)
        intent.putExtra(DisplayListActivity.INTENT_PLAY_FILE, bookName)
        val pendingIntent = PendingIntent.getActivity(
            context, 1, intent,
            PendingIntent.FLAG_CANCEL_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val style = NotificationCompat.MediaStyle()
            .setMediaSession(mediaSession.sessionToken)
            .setShowActionsInCompactView(1, 2, 3)
            .setShowCancelButton(true)
            .setCancelButtonIntent(
                MediaButtonReceiver.buildMediaButtonPendingIntent(
                    context,
                    PlaybackStateCompat.ACTION_STOP
                )
            )

        val notificationBuilder = androidx.core.app.NotificationCompat.Builder(context, CHANNEL_ID)
        notificationBuilder
            .setDeleteIntent(
                MediaButtonReceiver.buildMediaButtonPendingIntent(
                    context,
                    PlaybackStateCompat.ACTION_STOP
                )
            )
            .setVisibility(androidx.core.app.NotificationCompat.VISIBILITY_PUBLIC)
            .setSmallIcon(R.drawable.ic_logo)
            .setShowWhen(false)
            .setStyle(style)
            .setContentIntent(pendingIntent)
            .setContentTitle(description.title)
            .setContentText(description.subtitle)
            .setLargeIcon(description.iconBitmap)
            .setOngoing(playing)
        return notificationBuilder.build()
    }
}
